Add field stats api

martijnvg opened the pull request:
The field stats api should return field level statistics such as lowest, highest values and number of documents that have at least one value for a field.

An api like this can be useful to explore a data set you don't know much about. For example you can figure at with the lowest and highest response times are, so that you can create a histogram or range aggregation with sane settings.

This api doesn't run a search to figure this statistics out, but rather use the Lucene index look these statics up (using `Terms` class in Lucene). So finding out these stats for fields is cheap and quick.

The min/max values are based on the type of the field. So for a numeric field min/max are numbers and date field the min/max date and other fields the min/max are term based. 

The api is quite straight forward it expects a list of fields and optionally an index can be provided:
(ran on a stack exchange dump)

```
curl -XGET "http://localhost:9200/_field_stats?fields=comment_count,view_count,answer_count,creation_date,display_name"
```

The response looks like this:

```json
{
   "_shards": {
      "total": 1,
      "successful": 1,
      "failed": 0
   },
   "indices" : {
      "_all" : {
         "fields": {
            "comment_count": {
               "doc_count": 202940,
               "min_value": 0,
               "max_value": 98
            },
            "creation_date": {
               "doc_count": 564633,
               "min_value": "2008-08-01T16:37:51.513Z",
               "max_value": "2013-06-02T03:23:11.593Z"
            },
            "display_name": {
               "doc_count": 126741,
               "min_value": "0",
               "max_value": "정혜선"
            },
            "answer_count": {
               "doc_count": 139885,
               "min_value": 0,
               "max_value": 160
            },
            "view_count": {
               "doc_count": 152357,
               "min_value": 1,
               "max_value": 506955
            }
         }
      }
   }
}
```

This is a work in progress. Things to be done are documentation and tests.

rmuir commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -0,0 +1,311 @@
+        @Override
+        public void append(FieldStats stats) {
+            Lng other = (Lng) stats;
+            this.docCount += other.docCount;
```
This needs to handle the -1 case.

martijnvg replied:
good catch! I had this logic, refactored and forgot to add it back...

rmuir commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -0,0 +1,311 @@
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            docCount = in.readInt();
```
I don't think this should be an int if it gets summed up, it could exceed 2B?

martijnvg replied:
agreed, this should be a long

rmuir commented:
Should we expose the other stats in Terms too? I agree with leaving out size(), but these two are easily summed up just like docCount:
* sumDocFreq: number of postings
* sumTotalTermFreq: number of tokens

martijnvg commented:
@rmuir Thanks for taking a look at this! Agreed, it makes sense to expose `sumDocFreq ` and `sumTotalTermFreq` as well, since it is trivial to add those. Adding `size()` is far from trivial, since we need to have the actual terms at reduce time...

rmuir commented:
Should we return max_doc as well (i know it comes from the IndexReader)? The reason is, its needed for some calculations (like determining the density/sparseness of a field) and it seems screwed up if the user has to use another api just to compute that. e.g. `displayname` field is populated 54% of the time (doc_count/max_doc)

martijnvg commented:
> Should we return max_doc as well (i know it comes from the IndexReader)?

+1 Lets add that one too, it is very useful. I'll just pass `maxDoc` down with `Terms`.

martijnvg commented:
@rmuir I've updated the PR based on your feedback and added tests and docs.

rmuir commented:
Thanks for the update @martijnvg , a few questions:

For `field_ratio`, can we rename it? I think ratio is a little confusing (it makes me think of M:N). I would prefer something like `density`, and just document it is a percentage.

Along the same note, as `field_ratio` is a derived statistic, should we include other useful ones in a followup PR? Technically, that one a user could compute themselves, but I think its nice to have a plan.

Other potentially useful ones similar to that:

* `avg_length` (average number of words per doc): 

    sum_total_term_freq / doc_count

note: sum_total_term_freq will be -1 if frequencies are omitted, but in that case this one is easy, its 1 :)

* `avg_num_terms` (average number of unique words per doc): 

    sum_doc_freq / doc_count

note: both the above don't consider stopwords, they reflect what is indexed.
These can help you know how short/long the field is and so on.

Another interesting one:

* `multi_valued` (more than one indexed token in any doc): 

    sum_doc_freq != -1 && sum_doc_freq > doc_count

note: this currently is no good for numeric/date fields (will always say true, because of "extra tokens"), but works in the future with autoprefix.
This could maybe help with debugging fielddata or just understanding what is there.

I will take a look at the code tomorrow, thank you!

martijnvg commented:
> For field_ratio, can we rename it? I think ratio is a little confusing (it makes me think of M:N). I would prefer something like density, and just document it is a percentage.

+1, it was just to best name I could come up with.

I like the idea of adding more derived statistics. In fact because of these derived statistics and the good use for debugging purposes, I think it makes sense to add a _cat api variant of this api too? This api can nicely be rendering in a tabular format in the console.

> note: sum_total_term_freq will be -1 if frequencies are omitted, but in that case this one is easy, its 1 :)

Cool, that makes sense :)

> note: both the above don't consider stopwords, they reflect what is indexed. These can help you know how short/long the field is and so on.

I think if in the documentation we emphasised on the fact that all stats are based on indexed tokens, this is less of an surprise.

> note: this currently is no good for numeric/date fields (will always say true, because of "extra tokens"), but works in the future with autoprefix.

I like the `multi_valued` derived statistic. I'm just worried it will be misinterpret until autoprefix gets in, so maybe we should add that one later? (or at least not backport it to 1.6)

I also have been thinking about how to expose `Terms.size()` in a minimal way. I think it makes since to expose it as a per shard statistic, so it shows the number of terms for shard that holds the most terms? This way at least there is some insight in the number of terms per field.

rmuir commented:
> I like the multi_valued derived statistic. I'm just worried it will be misinterpret until autoprefix gets in, so maybe we should add that one later? (or at least not backport it to 1.6)

Well we already know the type of field, so i was thinking to just make it unavailable for numeric fields for now. But we'd still have it for strings.



> I also have been thinking about how to expose Terms.size() in a minimal way. I think it makes since to expose it as a per shard statistic, so it shows the number of terms for shard that holds the most terms? This way at least there is some insight in the number of terms per field.

I don't think we should do this. It would have to be the number of terms for a segment... Lets stay away from this one because the minute we go down this path, the API becomes slow.

martijnvg commented:
> Well we already know the type of field, so i was thinking to just make it unavailable for numeric fields for now. But we'd still have it for strings.

That makes sense.

> I don't think we should do this. It would have to be the number of terms for a segment... Lets stay away from this one because the minute we go down this path, the API becomes slow.

I totally missed that MultiTerms#size() isn't implemented, so yes doing that via the TermsEnum would be slow.

rmuir commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -0,0 +1,452 @@
+     */
+    public void append(FieldStats stats) {
+        this.maxDoc += stats.maxDoc;
+        if (stats.docCount != -1) {
```
maybe it is a matter of style, but i think its easier to handle the exceptional case like a guard up front: check stats.docCount == -1 and set to -1, otherwise sum. this is not really important to me.

rmuir commented:
changes look good to me. datatypes for all stats/ -1 handling / etc is all good. Maybe we want to rename the `field_ratio` -> `density` and push the change? We can open followups for the other stats ideas.

martijnvg commented:
I updated the PR and renamed `field_ratio` to `density`. But yes lets do the other derived measurements and _cat api variant in different issues.

rmuir commented on docs/reference/search/field-stats.asciidoc:
```diff
@@ -0,0 +1,101 @@
+
+* `max_doc`             - The total number of documents.
+* `doc_count`           - The number of documents that have at least one term for this field, or -1 if this measurement isn't available on one or more shards.
+* `density`             - A ratio as a percentage between all the documents that have at least one value for this field and all documents.
```
maybe we can reword the description when pushing, like "The percentage of documents that have at least one value for this field"

rmuir commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -0,0 +1,451 @@
+    }
+
+    /**
+     * @return A ratio as percentage of between all the documents that have at least one value for this field and all documents. (doc_count / max_doc)
```
and maybe drop any improved description here too

rmuir commented:
+1 (just some more nitpicks). thanks martijn, this is great.

rjernst commented on docs/reference/search/field-stats.asciidoc:
```diff
@@ -0,0 +1,101 @@
+
+experimental[]
+
+The field stats api allows to easily figure certain statistical properties without a field without executing a search, but
```
Maybe reword to:
`...allows one to find statistical properties of a field without...`

rjernst commented on docs/reference/search/field-stats.asciidoc:
```diff
@@ -0,0 +1,101 @@
+
+The field stats api allows to easily figure certain statistical properties without a field without executing a search, but
+looking up measurements that are natively available in the Lucene index. This can be useful to explore a data set,
+that you don't know much about in order to for example create histogram aggregation with a meaningful interval.
```
I would rework the second sentence into this:
`This can be useful to explore a dataset which you don't know much about. For example, this allows creating a histogram aggregation with meaningful intervals`

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -0,0 +1,451 @@
+
+import java.io.IOException;
+
+/**
```
remove empty javadoc

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -0,0 +1,451 @@
+    public void append(FieldStats stats) {
+        this.maxDoc += stats.maxDoc;
+        if (stats.docCount == -1) {
+            this.docCount = -1;
```
What if this.docCount was already -1? then it should stay -1 right?

martijnvg replied:
Yes, that is the case! I think only Lucene3x segments produce a `-1`, right? I wanted to write lucene3x segments in order to test this, but I can only test this in 1.x branch, since that is on Lucene4... Anyway I'll write a unit test for this method, I think that should cover it.

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStatsRequest.java:
```diff
@@ -0,0 +1,60 @@
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        if (fields != null) {
+            out.writeBoolean(true);
```
can you use `out.writeStringArrayNullable()` here instead?

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStatsShardRequest.java:
```diff
@@ -0,0 +1,70 @@
+        this.fields = request.fields();
+    }
+
+    public String[] fields() {
```
I know this pattern is used in other places, but can we stop the insanity and name this `getFields()` instead?

martijnvg replied:
++ I'm not a big fan of the pattern we use currently.

javanna replied:
I hate to be that guy...but this is user facing stuff through java API. I think we should use the same convention in all java api classes. If we want to use regular getters, which I would be very happy with, I think we should go and make the change on all existing classes consistently. Otherwise, doesn't it become confusing for users to have some classes one way and some other classes another way?

martijnvg replied:
@javanna FieldStatsShardRequest isn't user facing, but the FieldStatsRequest is. 

I don't like fields(String...), but your point is valid, so I'll make sure that FieldStatsRequest has this method.

At some point we should change the current getter/setter pattern on all request / response classes to the standard getter/setter pattern...

javanna replied:
Alright agreed, thanks Martijn.

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStatsShardRequest.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        if (fields != null) {
```
ditto on using `writeStringArrayNullable()`

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/TransportFieldStatsTransportAction.java:
```diff
@@ -0,0 +1,167 @@
+        int successfulShards = 0;
+        int failedShards = 0;
+        Map<String, FieldStats> mergedFieldStats = new HashMap<>();
+        List<ShardOperationFailedException> shardFailures = null;
```
Could you just initialize the array here and pass and so int he case of no failures an empty array would be passed to the response?

rjernst commented on src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java:
```diff
@@ -0,0 +1,119 @@
+                "float", "type=float", "long", "type=long", "integer", "type=integer", "short", "type=short", "byte", "type=byte"
+        ));
+
+        byte minByte = Byte.MAX_VALUE;
```
maybe make a separate function for testing each type? and without randomness? can these be tested with single node tests for the basics, then have the integ test to check the merging?

rjernst commented:
@martijnvg I left some comments, looks ok in general.  It is unfortunate so much boiler plate is necessary for a simple api :(

martijnvg commented:
@rjernst @rmuir  Thanks for the feedback. I'll update the PR later today. I agree adding a new api shouldn't add so much boilerplate code. This can definitely be done with less code. (Ideally we should have 3 classes per api: request, response and an operation class. I think for many broadcast apis this is doable. Hopefully this can done in the not so far future!)

rashidkpc commented:
Would it be possible to filter the fields/indices we get back based on a filter? Perhaps I want back all of the indices with `@timestamps` fields between 2014-01-01 and 2015-01-01:

```
curl -XGET "http://localhost:9200/_field_stats?fields=@timestamp&min=2014-01-01&max=2015-01-01"
```

This would be **incredibly** useful for Kibana and let us get rid of some serious bugs that occur for users with large numbers of indices. In fact, we could entirely get rid of our notion of timestamped indices as we'd be able to reliably sort an index list based on time. Users would only need to know their wildcard pattern! It would also allow for weird indexing strategies where maybe they're indexing weekly at one point and they need to step up to daily, something that isn't currently possible. Heck, you wouldn't even need timestamped indices, you could just increase a counter!

It would be so amazing, and so fast!

Martijn, I can't tell you how excited I am for this feature.

martijnvg commented:
@rashidkpc Yes, I think we can add this filtering on the min value and max value in the reduce phase of this api. So we effectively omit shard level results with a `min_value` and `max_value` that fall outside of the defined `min` and `max`. First I'll make sure that this gets pushed and I'll create a followup issue for this enhancement.

martijnvg commented:
I updated the PR to the latest master and addressed @rjernst's comments. (latest commit)

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -0,0 +1,451 @@
+        out.writeLong(sumTotalTermFreq);
+    }
+
+    public static class Lng extends FieldStats<Long> {
```
I don't think the short names actually buy anything? Could these have normal names like Long here?

martijnvg replied:
I like to rename it to `Long`, but then it clashes with `java.lang.Long`. Since these concrete classes are kind of private to ES the name shouldn't be an problem? I can also change the generic type to `java.lang.Long`.

rjernst replied:
Can't you just use FieldStats<java.lang.Long>? That is the only place the boxed type appears right?

rjernst replied:
Weird markdown seemed to silently remove some of my text...I was trying to say `FieldStats<java.lang.Long>` (which is what I think you meant by your last statement).

rjernst commented:
@martijnvg LGTM, I left a couple more minor comments. I also accidentally left most of them on the last commit itself...sorry.

martijnvg commented:
I updated the PR with the follow changes:
* Incorporated @rjernst feedback (thank you!)
* Added a `level` option to control whether the field stats of all indices are combined or field stats are only combined on a per index basis.
* Added better error handling in the case fields with the same are of a different type between indices.

martijnvg commented:
If there are no further comments about the additional commits then I'll push this tomorrow.

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/TransportFieldStatsTransportAction.java:
```diff
@@ -0,0 +1,186 @@
+                } else if ("indices".equals(request.level())) {
+                    indexName = shardResponse.getIndex();
+                } else {
+                    // should already have been catched by the FieldStatsRequest#validate(...)
```
catched -> caught

rjernst commented on src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -0,0 +1,455 @@
+
+    }
+
+    public static final class Text extends FieldStats<BytesRef> {
```
Should we call this `String` since that is what we call the field type?

rjernst commented:
@martijnvg LGTM, i just left a couple minor suggestions.

maharg101 commented:
Very nice. Will elasticsearch core use this internally to optimise which indices to query for a given field value ?

pjcard commented:
Am I right in thinking that anyone wishing to replace their time based indices with this mechanism will need to perform two queries in order to do so? (The first field_api query to obtain the indicies to pass in the URI for the search query). Is it possible to add this as a filter for a search query, thus doing away for the need use patterns in the search URI completely?

Or is @maharg101 correct in postulating that elasticsearch will use this internally to remove the need for the query to be explicit about indicies at all?

maharg101 commented:
@pjcard you have extrapolated what I meant slightly - the scenario I had in mind was that you would still be explicit about the time-based indices, and that the query execution would take advantage of the min / max values of fields to reduce the set of indices / shards to look in.

martijnvg commented:
@pjcard You would still need to send two requests to ES (1 field stats & 1 search request). ES doesn't use the min and max internally to optimise the search (all though in theory it can). This is also how Kibana is going to use the field stats api to reduce the number of indices a search needs to be executed on: https://github.com/elastic/kibana/issues/4342

pjcard commented:
Hi @martijnvg, thanks for the information. I'll add a request for that feature in case it's of use for anybody else and/or feasible.
